abdiff/core/calc_ab_diffs.py

```diff
@@ -52,29 +53,56 @@ def calc_ab_diffs(run_directory: str, collated_dataset_path: str) -> str:
     return str(diffs_dataset)
 
 
+def process_batch(batch: pa.RecordBatch) -> pa.RecordBatch:
+    """Parallel worker for calculating record diffs for a batch.
+
+    The pyarrow RecordBatch is converted into a pandas dataframe, a diff is calculated via
+    DeepDiff for each record in the batch, and this is converted back to a pyarrow
+    RecordBatch for returning.
+    """
+    df = batch.to_pandas()  # noqa: PD901
+    diff_results = df.apply(
+        lambda row: calc_record_diff(row["record_a"], row["record_b"]), axis=1
+    )
+    df["ab_diff"] = diff_results.apply(lambda x: x[0])
+    df["modified_timdex_fields"] = diff_results.apply(
+        lambda x: list(x[1]) if x[1] else []
+    )
+    df["has_diff"] = diff_results.apply(lambda x: x[2])
+    return pa.RecordBatch.from_pandas(df)  # type: ignore[attr-defined]
+
+
 def get_diffed_batches_iter(
     collated_dataset: ds.Dataset,
     batch_size: int = READ_BATCH_SIZE,
+    max_parallel_processes: int = MAX_PARALLEL_WORKERS,
 ) -> Generator[pa.RecordBatch, None, None]:
-    """Yield pyarrow record batches with diff calculated for records in batch."""
+    """Yield pyarrow record batches with diff calculated for each record.
+
+    This work is performed in parallel, leveraging CPU cores to calculate the diffs and
+    yield batches for writing to the "diffs" dataset.
+    """
     batches_iter = collated_dataset.to_batches(batch_size=batch_size)
-    for i, batch in enumerate(batches_iter):
-        logger.info(f"Calculating AB diff for batch: {i}")
-
-        # convert batch to pandas dataframe and calc values for new columns
-        df = batch.to_pandas()  # noqa: PD901
-
-        # calculate all diffs and unpack into separate columns
-        diff_results = df.apply(
-            lambda row: calc_record_diff(row["record_a"], row["record_b"]), axis=1
-        )
-        df["ab_diff"] = diff_results.apply(lambda x: x[0])
-        df["modified_timdex_fields"] = diff_results.apply(
-            lambda x: list(x[1]) if x[1] else []
-        )
-        df["has_diff"] = diff_results.apply(lambda x: x[2])
-
-        yield pa.RecordBatch.from_pandas(df)  # type: ignore[attr-defined]
+
+    with concurrent.futures.ProcessPoolExecutor(
+        max_workers=max_parallel_processes + 1
+    ) as executor:
+        pending_futures = []
+        for batch_count, batch in enumerate(batches_iter):
+            logger.info(f"Submitting batch {batch_count} for processing")
+            future = executor.submit(process_batch, batch)
+            pending_futures.append((batch_count, future))
+
+            if len(pending_futures) >= max_parallel_processes:
+                idx, completed_future = pending_futures.pop(0)
+                result = completed_future.result()
+                logger.info(f"Yielding diffed batch: {idx}")
+                yield result
```

Comment:
Side note: I learned something I had not considered about parallel workers during this work.

I had thought that `ProcessPoolExecutor(max_workers=X)` was sufficient to reduce CPU _and_ memory load, but that's only partially true.  While only `X` workers will actively use the CPU at once, they will be _scheduled_ all together at once, as will the arguments for each worker.  If the arguments -- data passed to the worker -- is large, this can actually put a large amount of data into memory before the worker even starts "working" on it.

As such, sometimes we need to actually throttle how quickly we even _schedule_ workers.  This code ensures that the "in flight" workers is less than a threshold, and we aren't scheduling too many at once.

Consider this example.  We have 100gb of data, but we'll only pass 1gb to each worker to perform calculations on, and we'll only have `ProcessPoolExecutor(max_workers=10)`.  I had thought this meant only 10 workers x 1gb = 10gb in memory at any given time.  But actually, we would _schedule_ 100 workers, each with 1gb of "arguments" (input data), and therefore 100gb could be stored in memory.  Sure, only 10 workers would be "working" on the 1gb they were passed each, but the waiting workers would still have their input data already in memory.  This is particularly sneaky when the input arguments are coming from an iterator; it would "drain" the iterator (the full 100gb) to schedule the workers.

What we want is actually only < 10 workers even scheduled at once, thereby slowly consuming the iterator of 100gb, and keeping memory pressure low.

This is probably poorly worded, and happy to discuss more, but worth noting.  FWIW, I have confirmed that this is memory safe as-is by following the pattern of throttling how many workers we have submitted to the `ProcessPoolExecutor()`.

Reply:
Update: I've made a personal note to revisit this.  I know for a fact that this works, and is memory safe at scale, so perhaps it's okay to proceed with it as-is at the moment.  But I'm not 100% sure what I've outlined above is accurate; I'm personally interested in digging into this a bit more, and confirming if workers can be scheduled with sizable amounts of data even before their processes are formally begun.